src/starknet_stub.rs

```diff
@@ -78,6 +78,191 @@ pub struct ContractLogs {
+
+    U256 { lo, hi }
+}
+
+pub fn encode_str_as_felts(msg: &str) -> Vec<Felt> {
+    const CHUNK_SIZE: usize = 32;
+
+    let data = msg.as_bytes().chunks(CHUNK_SIZE - 1);
+    let mut encoding = vec![Felt::default(); data.len()];
+    for (i, data_chunk) in data.enumerate() {
+        let mut chunk = [0_u8; CHUNK_SIZE];
+        chunk[1..data_chunk.len() + 1].copy_from_slice(data_chunk);
+        encoding[i] = Felt::from_bytes_be(&chunk);
+    }
+    encoding
+}
+
+pub fn decode_felts_as_str(encoding: &[Felt]) -> String {
+    let bytes_err: Vec<_> = encoding
+        .iter()
+        .flat_map(|felt| felt.to_bytes_be()[1..32].to_vec())
+        .collect();
+
+    match String::from_utf8(bytes_err) {
+        Ok(s) => s.trim_matches('\0').to_owned(),
+        Err(_) => {
+            let err_msgs = encoding
+                .iter()
+                .map(
+                    |felt| match String::from_utf8(felt.to_bytes_be()[1..32].to_vec()) {
+                        Ok(s) => format!("{} ({})", s.trim_matches('\0'), felt),
+                        Err(_) => felt.to_string(),
+                    },
+                )
+                .join(", ");
+            format!("[{}]", err_msgs)
+        }
+    }
+}
+
+impl<Curve: SWCurveConfig> Secp256Point<Curve>
+where
+    Curve::BaseField: PrimeField, // constraint for get_point_by_id
+{
+    // Given a (x,y) pair it will
+    // - return the point at infinity for (0,0)
+    // - Err if either x or y is outside of the modulus
+    // - Ok(None) if (x,y) are within the modules but not on the curve
+    // - Ok(Some(Point)) if (x,y) are on the curve
+    fn new(x: U256, y: U256) -> Result<Option<Self>, Vec<Felt>> {
+        let x = u256_to_biguint(x);
+        let y = u256_to_biguint(y);
+        let modulos = Curve::BaseField::MODULUS.into();
+
+        if x >= modulos || y >= modulos {
+            let error = Felt::from_hex(
+                "0x00000000000000000000000000000000496e76616c696420617267756d656e74",
+            ) // INVALID_ARGUMENT
+            .map_err(|err| encode_str_as_felts(&err.to_string()))?;
+
+            return Err(vec![error]);
+        }
+
+        Ok(maybe_affine(x.into(), y.into()))
+    }
+
+    fn add(p0: Self, p1: Self) -> Self {
+        let result: Projective<Curve> = p0.0 + p1.0;
+        Secp256Point(result.into())
+    }
+
+    fn mul(p: Self, m: U256) -> Self {
+        let result = p.0 * Curve::ScalarField::from(u256_to_biguint(m));
+        Secp256Point(result.into())
+    }
+
+    fn get_point_from_x(x: U256, y_parity: bool) -> Result<Option<Self>, Vec<Felt>> {
+        let modulos = Curve::BaseField::MODULUS.into();
+        let x = u256_to_biguint(x);
+
+        if x >= modulos {
+            let error = Felt::from_hex(
+                "0x00000000000000000000000000000000496e76616c696420617267756d656e74",
+            ) // INVALID_ARGUMENT
+            .map_err(|err| encode_str_as_felts(&err.to_string()))?;
```

Comment:
I think we have a function to convert a string into a felt.

Reply:
hmm i think i prefer to leave this as is, to match exactly the blockifier impl